File: common/src/main/java/juuxel/adorn/item/FuelData.java
package juuxel.adorn.item;

import juuxel.adorn.block.AdornBlocks;
import juuxel.adorn.lib.AdornTags;
import juuxel.adorn.lib.registry.Registered;
import net.minecraft.world.item.Item;
import net.minecraft.world.level.ItemLike;
import net.minecraft.world.item.ItemStack;
import net.minecraft.tags.TagKey;

import java.util.Set;

public sealed interface FuelData {
    Set<FuelData> FUEL_DATA = Set.of(
        // Wooden (300)
        new ForTag(AdornTags.CHAIRS.item(), 300),
        new ForTag(AdornTags.DRAWERS.item(), 300),
        new ForTag(AdornTags.TABLES.item(), 300),
        new ForTag(AdornTags.BENCHES.item(), 300),
        new ForTag(AdornTags.WOODEN_POSTS.item(), 300),
        new ForTag(AdornTags.WOODEN_PLATFORMS.item(), 300),
        new ForTag(AdornTags.WOODEN_STEPS.item(), 300),
        new ForTag(AdornTags.WOODEN_SHELVES.item(), 300),
        new ForItem(AdornBlocks.CRATE, 300),
        // Woollen (150)
        new ForTag(AdornTags.SOFAS.item(), 150)
    );

    int burnTime();
    boolean matches(ItemStack stack);

    record ForItem(Registered<? extends ItemLike> item, int burnTime) implements FuelData {
        @Override
        public boolean matches(ItemStack stack) {
            return stack.is(item.get().asItem());
        }
    }

    record ForTag(
        TagKey<Item> tag, int burnTime) implements FuelData {
        @Override
        public boolean matches(ItemStack stack) {
            return stack.is(tag);
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/item/FuelData.kt
package juuxel.adorn.item

import juuxel.adorn.block.AdornBlocks
import juuxel.adorn.lib.AdornTags
import net.minecraft.item.Item
import net.minecraft.item.ItemConvertible
import net.minecraft.item.ItemStack
import net.minecraft.tag.TagKey

sealed interface FuelData {
    val burnTime: Int
    fun matches(stack: ItemStack): Boolean

    data class ForItem(val item: ItemConvertible, override val burnTime: Int) : FuelData {
        override fun matches(stack: ItemStack): Boolean {
            return stack.isOf(item.asItem())
        }
    }

    data class ForTag(val tag: TagKey<Item>, override val burnTime: Int) : FuelData {
        override fun matches(stack: ItemStack): Boolean {
            return stack.isIn(tag)
        }
    }

    companion object {
        @JvmField
        val FUEL_DATA = setOf(
            // Wooden (300)
            ForTag(AdornTags.CHAIRS.item, 300),
            ForTag(AdornTags.DRAWERS.item, 300),
            ForTag(AdornTags.TABLES.item, 300),
            ForTag(AdornTags.BENCHES.item, 300),
            ForTag(AdornTags.WOODEN_POSTS.item, 300),
            ForTag(AdornTags.WOODEN_PLATFORMS.item, 300),
            ForTag(AdornTags.WOODEN_STEPS.item, 300),
            ForTag(AdornTags.WOODEN_SHELVES.item, 300),
            ForItem(AdornBlocks.CRATE, 300),
            // Woollen (150)
            ForTag(AdornTags.SOFAS.item, 150)
        )
    }
}
